### PR TITLE
fix `SearchAnchor` disposing `SearchController` while it is still used

### LINKED ISSUE #155180: SearchAnchor tries to use its controller after being disposed
### Steps to reproduce

1. Open the drawer.
2. Click on the search icon. The search view appears.
3. Double click outside of the search view AND outside the drawer as well. 

### Expected results

The first click should close the search view, while the second should close the drawer. This should happen without any errors.

### Actual results

The following exception is thrown:

```
════════ Exception caught by widgets library ═══════════════════════════════════
The following assertion was thrown building RawGestureDetector(state: RawGestureDetectorState#118c4(gestures: [tap, long press, tap and pan], excludeFromSemantics: true, behavior: translucent)):
A SearchController was used after being disposed.
Once you have called dispose() on a SearchController, it can no longer be used.

The relevant error-causing widget was:
    MaterialApp MaterialApp:file:///home/user/project/lib/main.dart:12:18

When the exception was thrown, this was the stack:
#0      ChangeNotifier.debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:183:9)
#1      ChangeNotifier.debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:190:6)
#2      ChangeNotifier.addListener (package:flutter/src/foundation/change_notifier.dart:275:27)
#3      _MergingListenable.addListener (package:flutter/src/foundation/change_notifier.dart:502:14)
#4      _AnimatedState.didUpdateWidget (package:flutter/src/widgets/transitions.dart:118:25)
#5      StatefulElement.update (package:flutter/src/widgets/framework.dart:5789:55)
#6      Element.updateChild (package:flutter/src/widgets/framework.dart:3941:15)
#7      SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6907:14)
#8      Element.updateChild (package:flutter/src/widgets/framework.dart:3941:15)
#9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5642:16)
#10     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5780:11)
#11     Element.rebuild (package:flutter/src/widgets/framework.dart:5333:7)
#12     StatefulElement.update (package:flutter/src/widgets/framework.dart:5803:5)
#13     Element.updateChild (package:flutter/src/widgets/framework.dart:3941:15)
#14     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5642:16)
#15     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5780:11)
#16     Element.rebuild (package:flutter/src/widgets/framework.dart:5333:7)
#17     StatefulElement.update (package:flutter/src/widgets/framework.dart:5803:5)
#18     Element.updateChild (package:flutter/src/widgets/framework.dart:3941:15)
#19     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6907:14)
#20     Element.updateChild (package:flutter/src/widgets/framework.dart:3941:15)
#21     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5642:16)
...
```

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: FirstScreen(),
    );
  }
}

class FirstScreen extends StatelessWidget {
  const FirstScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('The title')),
      drawer: Drawer(
        child: SearchAnchor(
          builder: (context, controller) {
            return Align(
              alignment: Alignment.topCenter,
              child: IconButton(
                onPressed: () => controller.openView(),
                icon: const Icon(Icons.search),
              ),
            );
          },
          suggestionsBuilder: (context, controller) {
            return [
              const ListTile(title: Text('Item 1')),
              const ListTile(title: Text('Item 2')),
              const ListTile(title: Text('Item 3')),
            ];
          },
        ),
      ),
      body: const Center(child: Text('Hi')),
    );
  }
}

```

</details>

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.3, on Ubuntu 22.04.4 LTS 6.8.0-40-generic, locale es_AR.UTF-8)
    • Flutter version 3.24.3 on channel stable at /home/fnicola/Programas/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (hace 2 días), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /home/fnicola/Android/Sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /home/fnicola/Programas/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2024.1)
    • Android Studio at /home/fnicola/Programas/android-studio
    • Flutter plugin version 81.0.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] VS Code (version 1.93.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.86.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.4 LTS 6.8.0-40-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 116.0.5845.140

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by PurplePolyhedron**
fixes https://github.com/flutter/flutter/issues/155180
New behaviour: SearchAnchor now closes the menu when itself is disposed while the menu is still open. This is the behaviour of `MenuAnchor`/`OverlayPortal`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/search_anchor.dart [hunk 1]
```diff
@@ -401,15 +402,26 @@ class _SearchAnchorState extends State<SearchAnchor> {
...
   @override
   void dispose() {
-    super.dispose();
     widget.searchController?._detach(this);
     _internalSearchController?._detach(this);
-    _internalSearchController?.dispose();
+    if (_route != null && _route!.isActive) {
+      _route!.navigator!.removeRoute(_route!);
+    }
+    // if the search route is still actice, the searchController would be
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/search_anchor.dart:410]:
```suggestion
    // If the search route is still active, the search controller would be
```

File: packages/flutter/lib/src/material/search_anchor.dart [hunk 2]
```diff
@@ -589,6 +603,11 @@ class _SearchViewRoute extends PopupRoute<_SearchViewRoute> {
   void dispose() {
     curvedAnimation?.dispose();
     viewFadeOnIntervalCurve?.dispose();
+    if (searchController._disposeByRoute) {
+      searchController.dispose();
+    }else {
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/search_anchor.dart:608]:
```suggestion
    } else {
```

File: packages/flutter/lib/src/material/search_anchor.dart [hunk 3]
```diff
@@ -1112,6 +1131,8 @@ class SearchController extends TextEditingController {
   // This is set automatically when a [SearchController] is given to the anchor
   // it controls.
   _SearchAnchorState? _anchor;
+  bool _usedByRoute = false;
+  bool _disposeByRoute = false;
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/search_anchor.dart:1135]:
```suggestion
  bool _disposedByRoute = false;
```

Just to be consistent.

**Comment by PurplePolyhedron** [comment]
I moved the disposal from `_SearchViewRoute` to `_ViewContentState` and added more comments.

**Review by nate-thegrate — COMMENTED** [review]
I'm really impressed with how the search anchor has been updated to elegantly handle any combination of the search view route & search anchor being created/disposed.

My main concern is the concept of "spaghetti code": adding a few extra members isn't a huge deal but does make the API slightly more difficult to understand & maintain.

<br>

My current inclination is to wait a few days to see if anyone has ideas for improving readability; if not, I'd be happy to approve this 🙂

File: packages/flutter/lib/src/material/search_anchor.dart [hunk 4]
```diff
@@ -371,6 +371,7 @@ class _SearchAnchorState extends State<SearchAnchor> {
   bool get _viewIsOpen => !_anchorIsVisible;
   SearchController? _internalSearchController;
   SearchController get _searchController => widget.searchController ?? (_internalSearchController ??= SearchController());
+  _SearchViewRoute? _route;
```
[the remark this answers is not in the document]
Comment (PurplePolyhedron) [reply packages/flutter/lib/src/material/search_anchor.dart:374]: The method is copied from `DropdownButton`, which is similar structure.

**Review by nate-thegrate — COMMENTED** [review]
Thanks for the updates!

<br>

I think I figured out a way to use 1 boolean flag instead of 2:

```dart
class SearchController {
  // If set to true, the SearchAnchor or ViewContent should
  // dispose of the controller when it's no longer in use.
  bool _internal = false;
}

class _SearchAnchorState {
  SearchController get _searchController {
    return widget.searchController
        ?? (_internalSearchController ??= SearchController().._internal = true);
  }

  void dispose() {
    widget.searchController?._detach(this);
    _internalSearchController?._detach(this);

    final _SearchViewRoute? route = _route;
    if (route != null && route.isActive) {
      route.navigator!.removeRoute(route);
    } else {
      _internalSearchController?.dispose();
    }
    super.dispose();
  }
}


class _ViewContentState {
  void dispose() {
    if (_controller._internal && !_controller.isAttached) {
      _controller.dispose();
    }
  }
}
```

Let me know what you think!

**Comment by PurplePolyhedron** [comment]
> Thanks for the updates!
> 
> 
> I think I figured out a way to use 1 boolean flag instead of 2:
> 
> ```dart
> class SearchController {
>   // If set to true, the SearchAnchor or ViewContent should
>   // dispose of the controller when it's no longer in use.
>   bool _internal = false;
> }
> 
> class _SearchAnchorState {
>   SearchController get _searchController {
>     return widget.searchController
>         ?? (_internalSearchController ??= SearchController().._internal = true);
>   }
> 
>   void dispose() {
>     widget.searchController?._detach(this);
>     _internalSearchController?._detach(this);
> 
>     final _SearchViewRoute? route = _route;
>     if (route != null && route.isActive) {
>       route.navigator!.removeRoute(route);
>     } else {
>       _internalSearchController?.dispose();
>     }
>     super.dispose();
>   }
> }
> 
> 
> class _ViewContentState {
>   void dispose() {
>     if (_controller._internal && !_controller.isAttached) {
>       _controller.dispose();
>     }
>   }
> }
> ```
> 
> Let me know what you think!

Thanks for the suggestion, I tried it but it crashes using the example from the issue. My bad for not covering that in the test cases.
I think this is due to the route is not considered `isActive` during the popping animation.

However it did inspired me to come up with a new idea.

File: packages/flutter/lib/src/material/search_anchor.dart [hunk 5]
```diff
@@ -401,15 +402,17 @@ class _SearchAnchorState extends State<SearchAnchor> {
...
   @override
   void dispose() {
-    super.dispose();
     widget.searchController?._detach(this);
     _internalSearchController?._detach(this);
-    _internalSearchController?.dispose();
-  }
+    _route?._dismiss(
+      disposeController: widget.searchController == null,
+    );
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/search_anchor.dart:409]:
```suggestion
    final bool usingExternalController = widget.searchController != null;
    _route?._dismiss(disposeController: !usingExternalController);
    if (usingExternalController) {
      _internalSearchController?.dispose();
    }
```

**Review by nate-thegrate — COMMENTED** [review]
This looks fantastic—I like this current setup a lot more than anything suggested previously!

I believe only one more change is needed:

```dart
  @override
  void didUpdateWidget(SearchAnchor oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.searchController != widget.searchController) {
      oldWidget.searchController?._detach(this);
      _searchController._attach(this);
    }
  }
```

Based on that method, it looks like there's an edge case where the SearchAnchor starts using the internal search controller, but later switches over to an external one—in that case, the internal controller would need to be disposed of.

File: packages/flutter/lib/src/material/search_anchor.dart [hunk 6]
```diff
@@ -401,15 +402,25 @@ class _SearchAnchorState extends State<SearchAnchor> {
...
     widget.searchController?._detach(this);
     _internalSearchController?._detach(this);
-    _internalSearchController?.dispose();
-  }
+    final bool usingExternalController = widget.searchController != null;
+    if (_viewIsOpen) {
+      _route?._dismiss(
+        disposeController: !usingExternalController,
+      );
+      if (usingExternalController){
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/search_anchor.dart:412]:
```suggestion
      if (usingExternalController) {
```

**Review by nate-thegrate — APPROVED** [review]
Flutter_LGTM

Thanks for the high-quality contribution!

File: packages/flutter/lib/src/material/search_anchor.dart (hunk 6 above)
Comment (PurplePolyhedron) [reply packages/flutter/lib/src/material/search_anchor.dart]: Thanks for the review and suggestions. This issue was quite tricky.

**Review by QuncCccccc — COMMENTED** [review]
Thanks for the fix! I haven't taken a closer look at this implementation but seems the sample code in the issue still throw the same exception:)

File: packages/flutter/test/material/search_anchor_test.dart [hunk 7]
```diff
@@ -3419,6 +3419,151 @@ void main() {
       expect(find.byType(Placeholder), findsOneWidget);
     });
   });
+
+  testWidgets('SearchAnchor does not dispose external SeachController', (WidgetTester tester) async {
```
Comment (QuncCccccc) [inline packages/flutter/test/material/search_anchor_test.dart:3423]:
```suggestion
  testWidgets('SearchAnchor does not dispose external SearchController', (WidgetTester tester) async {
```

**Review by QuncCccccc — APPROVED** [review]
LGTM! Thanks a lot for the fix!

**Pull request merged**